src/libstore/aws-creds.cc

```diff
@@ -0,0 +1,219 @@
+
+} // anonymous namespace
+
+AwsCredentials getAwsCredentials(const std::string & profile)
+{
+    // Get or create credential provider with caching
+    std::shared_ptr<Aws::Crt::Auth::ICredentialsProvider> provider;
+
+    // Try to find existing provider
+    credentialProviderCache.visit(profile, [&](const auto & pair) { provider = pair.second; });
+
+    if (!provider) {
+        // Create new provider if not found
+        debug(
+            "[pid=%d] creating new AWS credential provider for profile '%s'",
+            getpid(),
+            profile.empty() ? "(default)" : profile.c_str());
+
+        try {
+            initAwsCrt();
+
+            if (profile.empty()) {
+                Aws::Crt::Auth::CredentialsProviderChainDefaultConfig config;
+                config.Bootstrap = Aws::Crt::ApiHandle::GetOrCreateStaticDefaultClientBootstrap();
+                provider = Aws::Crt::Auth::CredentialsProvider::CreateCredentialsProviderChainDefault(config);
+            } else {
+                Aws::Crt::Auth::CredentialsProviderProfileConfig config;
+                config.Bootstrap = Aws::Crt::ApiHandle::GetOrCreateStaticDefaultClientBootstrap();
+                config.ProfileNameOverride = Aws::Crt::ByteCursorFromCString(profile.c_str());
+                provider = Aws::Crt::Auth::CredentialsProvider::CreateCredentialsProviderProfile(config);
+            }
+        } catch (Error & e) {
+            e.addTrace(
+                {},
+                "while creating AWS credentials provider for %s",
+                profile.empty() ? "default profile" : fmt("profile '%s'", profile));
+            throw;
+        }
+
+        if (!provider) {
+            throw AwsAuthError(
+                "Failed to create AWS credentials provider for %s",
+                profile.empty() ? "default profile" : fmt("profile '%s'", profile));
+        }
+
+        // Insert into cache (try_emplace is thread-safe and won't overwrite if another thread added it)
+        credentialProviderCache.try_emplace(profile, provider);
+    }
+
+    return getCredentialsFromProvider(provider);
+}
+
+void invalidateAwsCredentials(const std::string & profile)
+{
+    credentialProviderCache.erase(profile);
+}
+
+void clearAwsCredentialsCache()
+{
+    credentialProviderCache.clear();
+}
+
+std::optional<AwsCredentials> preResolveAwsCredentials(const std::string & url)
+{
+    try {
+        auto parsedUrl = parseURL(url);
+        if (parsedUrl.scheme != "s3") {
+            return std::nullopt;
+        }
+
+        auto s3Url = ParsedS3URL::parse(parsedUrl);
+        std::string profile = s3Url.profile.value_or("");
+
+        // Get credentials (automatically cached)
+        return getAwsCredentials(profile);
+    } catch (const AwsAuthError & e) {
+        debug("Failed to pre-resolve AWS credentials: %s", e.what());
+        return std::nullopt;
+    } catch (const std::exception & e) {
+        debug("Failed to pre-resolve AWS credentials: %s", e.what());
```

Comment:
I would always make the debug strings different so you can 1:1 map them when you see them to line numbers.